src/sparse/csmat.rs

```diff
@@ -2953,3 +2953,298 @@ mod test {
         assert_eq!(CsMat::eye(2), onehot);
     }
 }
+
+#[cfg(feature = "approx")]
+mod approx_impls {
+    use super::*;
+    use approx::*;
+
+    impl<N, I, Iptr, IS1, DS1, ISptr1, IS2, ISptr2, DS2>
+        AbsDiffEq<CsMatBase<N, I, ISptr2, IS2, DS2, Iptr>>
+        for CsMatBase<N, I, ISptr1, IS1, DS1, Iptr>
+    where
+        I: SpIndex,
+        Iptr: SpIndex,
+        CsMatBase<N, I, ISptr1, IS1, DS1, Iptr>:
+            std::cmp::PartialEq<CsMatBase<N, I, ISptr2, IS2, DS2, Iptr>>,
+        IS1: Deref<Target = [I]>,
+        IS2: Deref<Target = [I]>,
+        ISptr1: Deref<Target = [Iptr]>,
+        ISptr2: Deref<Target = [Iptr]>,
+        DS1: Deref<Target = [N]>,
+        DS2: Deref<Target = [N]>,
+        N: AbsDiffEq,
+        N::Epsilon: Clone,
+        N: num_traits::Zero,
+    {
+        type Epsilon = N::Epsilon;
+        fn default_epsilon() -> N::Epsilon {
+            N::default_epsilon()
+        }
+        fn abs_diff_eq(
+            &self,
+            other: &CsMatBase<N, I, ISptr2, IS2, DS2, Iptr>,
+            epsilon: N::Epsilon,
+        ) -> bool {
+            if self.shape() != other.shape() {
+                return false;
+            }
+            if self.storage() == other.storage() {
+                self.outer_iterator()
+                    .zip(other.outer_iterator())
+                    .all(|(r1, r2)| r1.abs_diff_eq(&r2, epsilon.clone()))
+            } else {
+                // Checks if all elements in self has a matching element
+                // in other
+                if !self.iter().all(|(n, (i, j))| {
+                    n.abs_diff_eq(
+                        other
+                            .get(i.to_usize().unwrap(), j.to_usize().unwrap())
+                            .unwrap_or(&N::zero()),
+                        epsilon.clone(),
+                    )
```

Comment:
I'm wondering if all these binary searches are not going to end up quite expensive, maybe in some cases it's best to convert one operand to the other's storage. But I'm not sure, since it may be costly to allocate a new matrix in a comparison.

Reply:
I tried the cloning approach, but that leads to problems with expected type parameters, seems one can't expect a recursive call to resolve to different parameters

Reply:
We could add a special case when all the elements in both are coinciding, so we only have to check the matrices one way instead of both

Reply:
The more I think about it, the more I think it's not a problem. It's ok for the comparison to be slower when the storages differ, the caller can deal with it by explictly changing storages, we simply need to document it.